lib/network.go

```diff
@@ -2018,13 +2078,104 @@ func (msg *MsgDeSoHeader) EncodeHeaderVersion1(preSignature bool) ([]byte, error
+	if transactionMerkleRoot == nil {
+		transactionMerkleRoot = &BlockHash{}
+	}
+	retBytes = append(retBytes, transactionMerkleRoot[:]...)
+
+	// TstampSecs: this field can be encoded to take up to the full 64 bits now
+	// that MsgDeSoHeader version 2 does not need to be backwards compatible.
+	retBytes = append(retBytes, UintToBuf(msg.TstampSecs)...)
+
+	// Height: similar to the field above, this field can be encoded to take
+	// up to the full 64 bits now that MsgDeSoHeader version 2 does not need to
+	// be backwards compatible.
+	retBytes = append(retBytes, UintToBuf(msg.Height)...)
+
+	// The Nonce and ExtraNonce fields are unused in version 2. We skip them
+	// during both encoding and decoding.
+
+	// ProposerPublicKey
+	if msg.ProposerPublicKey == nil {
+		return nil, fmt.Errorf("EncodeHeaderVersion2: ProposerPublicKey must be non-nil")
+	}
+	retBytes = append(retBytes, msg.ProposerPublicKey.ToBytes()...)
+
+	// ProposerVotingPublicKey
+	if msg.ProposerVotingPublicKey == nil {
+		return nil, fmt.Errorf("EncodeHeaderVersion2: ProposerVotingPublicKey must be non-nil")
+	}
+	retBytes = append(retBytes, EncodeBLSPublicKey(msg.ProposerVotingPublicKey)...)
+
+	// ProposedInView
+	retBytes = append(retBytes, UintToBuf(msg.ProposedInView)...)
+
+	// ValidatorsVoteQC
+	if msg.ValidatorsVoteQC == nil {
+		return nil, fmt.Errorf("EncodeHeaderVersion2: ValidatorsVoteQC must be non-nil")
+	}
+	encodedValidatorsVoteQC, err := msg.ValidatorsVoteQC.ToBytes()
+	if err != nil {
+		return nil, errors.Wrapf(err, "EncodeHeaderVersion2: error encoding ValidatorsVoteQC")
+	}
+	retBytes = append(retBytes, encodedValidatorsVoteQC...)
+
+	// ValidatorsTimeoutAggregateQC
+	if msg.ValidatorsTimeoutAggregateQC == nil {
+		return nil, fmt.Errorf("EncodeHeaderVersion2: ValidatorsTimeoutAggregateQC must be non-nil")
+	}
+	encodedValidatorsTimeoutAggregateQC, err := msg.ValidatorsTimeoutAggregateQC.ToBytes()
+	if err != nil {
+		return nil, errors.Wrapf(err, "EncodeHeaderVersion2: error encoding ValidatorsTimeoutAggregateQC")
+	}
+	retBytes = append(retBytes, encodedValidatorsTimeoutAggregateQC...)
+
+	// ProposerVotePartialSignature: we encode the signature if it's present and the preSignature
+	// flag is set to false. Otherwise, we encode an empty byte array as a placeholder. The placeholder
+	// ensures, that the DecodeHeaderVersion2 function can properly recognize encoding where a signature
+	// isn't populated. It ensures that every possible output from EncodeHeaderVersion2 can be decoded by
+	// DecodeHeaderVersion2.
+	proposerSignatureBytes := []byte{}
+	if !preSignature {
+		// If the the preSignature flag is set to false, then the caller intends to encode the signature.
+		// The signature must be non-nil.
+		if msg.ProposerVotePartialSignature == nil {
+			return nil, fmt.Errorf("EncodeHeaderVersion2: ProposerVotePartialSignature must be non-nil")
+		}
+		proposerSignatureBytes = msg.ProposerVotePartialSignature.ToBytes()
+	}
```

Comment:
Noting that we need to encode/decode the new fields here.

Reply:
I've added a new `EncodeOptionalBLSSignature(sig)` function for this